kroxylicious-kms-provider-aws-kms/src/test/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2MetadataCredentialsProviderTest.java

```diff
@@ -0,0 +1,320 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.github.tomakehurst.wiremock.WireMockServer;
+import com.github.tomakehurst.wiremock.client.WireMock;
+
+import io.kroxylicious.kms.provider.aws.kms.config.Ec2MetadataCredentialsProviderConfig;
+import io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProvider.SecurityCredentials;
+import io.kroxylicious.kms.service.KmsException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.put;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.wireMockConfig;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class Ec2MetadataCredentialsProviderTest {
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().registerModule(new JavaTimeModule());
+
+    private static final String MY_TOKEN = "mytoken";
+    private static final String IAM_ROLE = "myrole";
+    private static final String TOKEN_RETRIEVAL_ENDPOINT = "/latest/api/token";
+    private static final String META_DATA_IAM_SECURITY_CREDENTIALS_ENDPOINT = "/2024-04-11/meta-data/iam/security-credentials/";
+
+    // From https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-metadata-security-credentials.html
+    private static final String KNOWN_GOOD_SECURITY_CREDENTIAL_RESPONSE = """
+            {
+              "Code" : "Success",
+              "LastUpdated" : "2012-04-26T16:39:16Z",
+              "Type" : "AWS-HMAC",
+              "AccessKeyId" : "ASIAIOSFODNN7EXAMPLE",
+              "SecretAccessKey" : "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
+              "Token" : "token",
+              "Expiration" : "2017-05-17T15:09:54Z"
+            }""";
+    private static WireMockServer metadataServer;
+    private Ec2MetadataCredentialsProviderConfig config;
+    private ScheduledExecutorService executorService;
+
+    @BeforeAll
+    public static void initMockRegistry() {
+        metadataServer = new WireMockServer(wireMockConfig().dynamicPort());
+        metadataServer.start();
+    }
+
+    @AfterAll
+    public static void shutdownMockRegistry() {
+        metadataServer.shutdown();
+    }
+
+    @BeforeEach
+    void setUp() {
+        config = new Ec2MetadataCredentialsProviderConfig(IAM_ROLE, Optional.of(URI.create(metadataServer.baseUrl())), Optional.of(0.20));
+        executorService = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Does this need to be re-initalised for every test?

Reply:
I value knowing that the executor is in a clean state before each run. 

I think the test ought to be asserting that there are no pending runnables left on the executor.  The implementation ought really cancel scheduled tasks on close (which it is not doing right now).

Reply:
Actually, I realise I don't need the executor exposed directly to the test at all.